### ISSUE TITLE
sbJson 2.0 relatedItems

### ISSUE THREAD

**Issue by stansmith907** [adiwg/mdTranslator#81]
see issue #58

**Comment by stansmith907** [comment]
sbJson example below is updated from issue #58.   *relatedItems* is now parent to *link* when link type is "related".

````json
   "link": {
      "rel": "self",
      "url": "https://www.sciencebase.gov/catalog/item/57d97341e4b090824ffb0e6f"
   },
   "relatedItems": {
      "link": {
         "url": "https://www.sciencebase.gov/catalog/itemLinks?itemId=57d97341e4b090824ffb0e6f",
         "rel": "related"
      }
   }
````
To retrieve information about the the *relatedItems* access the URL with attribute format=json.  

````json
[
   {
      "id": "58e69c6ce4b09da6799ac898",
      "itemId": "57d97341e4b090824ffb0e6f",
      "relatedItemId": "5519b3f8e4b032384278330b",
      "itemLinkTypeId": "53d173c8e4b0536257c3415b",
      "relatedItemTitle": "An Assessment of Midwestern ...",
      "rel": "produced",
      "title": "Spatial data: A large-scale database of ...",
      "itemLinkTypePath": "/itemLink/type/productOf",
      "type": "productOf",
      "typeLabel": "product of",
      "typeReverseDescription": "produced"
   }
]
````
For sbJson READER >>>
Map each of the returned items to an 'associatedResource'.  

````ruby
   def newAssociatedResource
      {
         resourceTypes: [],
         associationType: nil,
         initiativeType: nil,
         resourceCitation: {
            title: "title",
            alternateTitles: [],
            dates: [],
            edition: nil,
            responsibleParties: [],
            presentationForms: [],
            identifiers: [
              {
                  identifier: "id",
                  namespace: "gov.sciencebase.catalog",
                  version: nil,
                  description: "identifier",
                 citation: {}
               },
              {
                  identifier: "itemId",
                  namespace: "gov.sciencebase.catalog",
                  version: nil,
                  description: "item ID",
                 citation: {}
               },
              {
                  identifier: "relatedItemId",
                  namespace: "gov.sciencebase.catalog",
                  version: nil,
                  description: "related item ID",
                 citation: {}
               },
              {
                  identifier: "itemLinkTypeId",
                  namespace: "gov.sciencebase.catalog",
                  version: nil,
                  description: "item link type ID",
                 citation: {}
               },
            ],
            series: {},
            otherDetails: [],
            onlineResources: [
               {
                  olResURI: "https://www.sciencebase.gov/catalog/item/{itemId}",
                  olResProtocol: nil,
                  olResName: nil,
                  olResDesc: nil,
                  olResFunction: nil
               }
            ],
            browseGraphics: []
         },
         metadataCitation: {}
      }
   end
````

**Comment by stansmith907** [comment]
We need to meet, likely with Lei Ann, to decide how to map in 'associatedResource' required elements of 'resourceType" and 'associationType', also optional element 'initiativeType'.

**Comment by jlblcc** [comment]
Link to SB vocab: https://www.sciencebase.gov/vocab/vocabulary/4f4e475de4b07f02db47decc

**Comment by stansmith907** [comment]
We need to clarify some of the associationType codes and definitions before completing this issue.  See issue 17 in mdCodes.

**Comment by stansmith907** [comment]
Here's the current associationType code mapping from-to ISO-sbJson:

````ruby
               @iso_sb_association = [
                  {iso: 'crossReference', sb: 'alternate'},
                  {iso: 'largerWorkCitation', sb: nil},
                  {iso: 'partOfSeamlessDatabase', sb: nil},
                  {iso: 'source', sb: nil},
                  {iso: 'stereoMate', sb: nil},
                  {iso: 'isComposedOf', sb: 'constituentOf'},
                  {iso: 'collectiveTitle', sb: nil},
                  {iso: 'series', sb: nil},
                  {iso: 'dependency', sb: nil},
                  {iso: 'revisionOf', sb: nil},
                  {iso: 'projectProduct', sb: nil, deprecated: true},
                  {iso: 'supplementalResource', sb: nil},
                  {iso: 'produced', sb: nil, deprecated: true},
                  {iso: 'productOf', sb: nil, deprecated: true},
                  {iso: 'mainProjectOf', sb: nil, deprecated: true},
                  {iso: 'subProjectOf', sb: nil, deprecated: true},
                  {iso: 'product', sb: 'derivativeOf'},
                  {iso: 'parentProject', sb: 'productOf'},
                  {iso: 'subProject', sb: 'subprojectOf'},
                  {iso: nil, sb: 'related'},
                  {iso: nil, sb: 'copiedFrom'},
                  {iso: nil, sb: 'precededBy'}
               ]
````

**Comment by stansmith907** [comment]
I will ignore the sciencebase ids on read as they are recreated in the intake process of sbJson records.

**Comment by jlblcc** [comment]
Ok, this is kinda screwed up, but that link to the vocab does not list all of the association types. To get the "reverse" association you have to look at the actual term. Example: https://www.sciencebase.gov/vocab/term/53d173c8e4b0536257c3415b

I'm just the messenger :gun:

**Comment by jlblcc** [comment]
So _product_ should map to _produced_, I think. Not sure about others, that's just the one I'm most familiar with.

**Comment by stansmith907** [comment]
Yeah, that's what I did.  But some sbJSON codes don't even have meager definitions.  Also, I deprecated the ADIwg code *produced* as it was a verb and replaced it with *product*.  

Before today *product* was neither an ADIwg or sbJSON code.  It is one of the new ones I added to ADIwg.  The closest sbJSON equivalent I found was *derivativeOf*.

**Comment by jlblcc** [comment]
Something **must** map to _produced_. _produced_ and _productOf_ are the associations used to define project:product relationships in ScienceBase.

**Comment by stansmith907** [comment]
I'm lost. *produced* is not on there code list.  Is this a secret code!?  *parentProject* maps to *productOf*.  Unless you see the sbJson codes as reversed.

**Comment by jlblcc** [comment]
Look at the link below. The "reverse description" is essentially used as another association type. We'll need to support that if we're going to do "two-way" mapping.

> Ok, this is kinda screwed up, but that link to the vocab does not list all of the association types. To get the "reverse" association you have to look at the actual term. Example: https://www.sciencebase.gov/vocab/term/53d173c8e4b0536257c3415b

Here's an example:

**Comment by jlblcc** [comment]
Here's the UI, notice what happens when I check the box:
![2017-08-02 16_41_19-associate a related item - sciencebase-catalog](https://user-images.githubusercontent.com/5833603/28900983-a4bb578c-77a1-11e7-9438-39d68c7dcad8.png)

![image](https://user-images.githubusercontent.com/5833603/28900977-993aa3d6-77a1-11e7-8b66-3deaee54e2e9.png)

**Comment by jlblcc** [comment]
Hmmmm.....in the end, SB only uses "productOf" as the "type" regardless of the relationship direction:
```
[
{
id: "597b8ca8e4b0a38ca27565a0",
itemId: "590b7498e4b0e541a0378cb5",
relatedItemId: "5764a757e4b07657d19ba995",
itemLinkTypeId: "53d173c8e4b0536257c3415b",
relatedItemTitle: "TEON: Terrestrial Environmental Observation Network",
rel: "produced",
title: "test",
itemLinkTypePath: "/itemLink/type/productOf",
type: "productOf",
typeLabel: "product of",
annotation: "",
typeReverseDescription: "produced"
},
{
id: "598271c4e4b0e2f5d463bbac",
itemId: "5764a757e4b07657d19ba995",
relatedItemId: "590b7498e4b0e541a0378cb5",
itemLinkTypeId: "53d173c8e4b0536257c3415b",
relatedItemTitle: "test",
rel: "produced",
title: "TEON: Terrestrial Environmental Observation Network",
itemLinkTypePath: "/itemLink/type/productOf",
type: "productOf",
typeLabel: "product of",
annotation: "",
typeReverseDescription: "produced"
}
]
```
We need to be able define the relationship from either end in mdJSON, but I guess we don't really need to map to _produced_ since we're not actually writing the associations in sbJSON. The reader will have to be smart enough to "reverse" the relationships based on the id values.

Clear as mud?

**Comment by jlblcc** [comment]
BTW, currently the SB mdJSON harvester expects "produced" and "productOf" to be used in the mdJSON. If we change those, they'll have to update the harvester.

**Comment by stansmith907** [comment]
Notes on mapping sbJson reversed relationships: 
* These descriptions are in the direction of how the relatedItem (associatedResource) relates to the main item (main resource).  
* I'm not sure the following descriptions were the effort.  In the above examples direction of the relation is not determined by a uniform pattern for relateditems or the codes.
* The relationship direction is determined by the IDs in itemId and relatedItemId.  These IDs must be matched to the main sbJson record id: to determine direction.  A match to itemId: is forward (same as ISO associationType) and a match to relatedItemId: is reversed.  However rel: and typeReverseDescription: always contain the reverse relationship code and type: always contains the forward code. 
* In a forward relationship title: is the name of the relatedItem (associatedResource) and relatedItemTitle: is the name of the main item.  
* In a reverse relationship title: is the name of main item and relatedItemTitle: is the name of the relatedItem. 

#### *copiedFrom*
If *copiedFrom* identifies the relatedItem is the source of the main item, *copiedInto* is just the main item.

#### *constituentOf* 
If *constituentOf* states the related item is only a part of main item, *constituent* seems to mean exactly the same thing, not actually the reverse.  We can make a call that *constituent* is the main item. 

#### *derivativeOf* 
If *derivativeOf* states the related item is derived from the main item, *derived* seems to mean exactly the same thing and is not actually the reverse.  *derived* is just the past tense verb for *derivativeOf*.

#### *precededBy*
*precededBy* has no definition and seems like an association for a linage process steps in some derivation.  And so would be the reverse *succeededBy*.  We are not mapping linage into assiciatedResource.  Possibly map this to lineage in the future?

#### *productOf* 
If *productOf* identifies the related item is the parent project for the main item, *produced* is just the main resource.

#### *related* 
The reverse of *related* is *related* so...

#### *subprojectOf* 
If *subprojectOf* identifies the related item is the parent project for the main item, *mainprojectOf* is just the main resource.

**Comment by stansmith907** [comment]
Note from Dell after today's LCC call 
----
      "id": "58e69c6ce4b09da6799ac898", //id for the itemLink?
MongoDB identifier for the item link object itself
 
      "itemId": "57d97341e4b090824ffb0e6f", //id for the item?
MongoDB identifier for the "from" SB catalog item
 
      "relatedItemId": "5519b3f8e4b032384278330b", //id of the relatedItem
MongoDB identifier for the "to" SB catalog item

      "itemLinkTypeId": "53d173c8e4b0536257c3415b", //id of the linkType vocab?
Vocab term ID for the link type, e.g. https://www.sciencebase.gov/vocab/term/53d173c8e4b0536257c3415b
 
      "relatedItemTitle": "An Assessment of Midwestern ...",
Title of the "to" item
 
      "rel": "produced", //reverse linkType?
The relationship description
 
      "title": "Spatial data: A large-scale database of ...",
Title of the "from" item
 
      "itemLinkTypePath": "/itemLink/type/productOf", //vocab path?
Vocab scheme path
 
      "type": "productOf", //type of link(relationship)?
Vocab link type name
 
      "typeLabel": "product of",
Link description to use on "to" item
 
      "typeReverseDescription": "produced" // same as "rel" or label for "rel"?

**Comment by stansmith907** [comment]
I think this is closed now, I made a few more comments on #137
